File: backend/app/schemas/diet_plan.py
import uuid
from datetime import date

from pydantic import BaseModel

from app.models.training_plan import DayOfWeek, PlanStatus
from app.schemas.meal_template import MacroTotals
from app.schemas.menu import MenuSummaryOut


class DietPlanCreate(BaseModel):
    title: str
    notes: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    status: PlanStatus = PlanStatus.DRAFT
    daily_calories_target: float | None = None
    daily_protein_g: float | None = None
    daily_carbs_g: float | None = None
    daily_fat_g: float | None = None


class DietPlanUpdate(BaseModel):
    title: str | None = None
    notes: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    status: PlanStatus | None = None
    daily_calories_target: float | None = None
    daily_protein_g: float | None = None
    daily_carbs_g: float | None = None
    daily_fat_g: float | None = None


class DietPlanOut(BaseModel):
    id: uuid.UUID
    client_id: uuid.UUID
    title: str
    notes: str | None
    start_date: date | None
    end_date: date | None
    status: PlanStatus
    daily_calories_target: float | None
    daily_protein_g: float | None
    daily_carbs_g: float | None
    daily_fat_g: float | None

    model_config = {"from_attributes": True}


class DietWeekCreate(BaseModel):
    week_number: int
    notes: str | None = None


class DietWeekOut(BaseModel):
    id: uuid.UUID
    week_number: int
    notes: str | None

    model_config = {"from_attributes": True}


class DietDayIn(BaseModel):
    day_of_week: DayOfWeek
    order_index: int
    menu_id: uuid.UUID | None = None


class DietDayOut(BaseModel):
    id: uuid.UUID
    day_of_week: DayOfWeek
    order_index: int
    menu: MenuSummaryOut | None
    totals: MacroTotals | None


class DietWeekDetailOut(DietWeekOut):
    days: list[DietDayOut]


class SetDietDaysRequest(BaseModel):
    days: list[DietDayIn]


class AssignMenuRequest(BaseModel):
    menu_id: uuid.UUID
    # None applies the menu to all 7 days of the week.
    days_of_week: list[DayOfWeek] | None = None


class DuplicateWeekRequest(BaseModel):
    week_number: int


class DietPlanDetailOut(DietPlanOut):
    weeks: list[DietWeekDetailOut]

File: frontend/src/types/diet-plan.ts
import type { DayOfWeek, PlanStatus } from "./common";
import type { MacroTotals, MenuSummary } from "./diet";

export interface DietPlan {
  id: string;
  client_id: string;
  title: string;
  notes: string | null;
  start_date: string | null;
  end_date: string | null;
  status: PlanStatus;
  daily_calories_target: number | null;
  daily_protein_g: number | null;
  daily_carbs_g: number | null;
  daily_fat_g: number | null;
}

export interface DietDay {
  id: string;
  day_of_week: DayOfWeek;
  order_index: number;
  menu: MenuSummary | null;
  totals: MacroTotals | null;
}

export interface DietWeek {
  id: string;
  week_number: number;
  notes: string | null;
  days: DietDay[];
}

export interface DietPlanDetail extends DietPlan {
  weeks: DietWeek[];
}

export interface DietPlanInput {
  title: string;
  notes?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  status?: PlanStatus;
  daily_calories_target?: number | null;
  daily_protein_g?: number | null;
  daily_carbs_g?: number | null;
  daily_fat_g?: number | null;
}

export interface DietDayInput {
  day_of_week: DayOfWeek;
  order_index: number;
  menu_id: string | null;
}

export interface AssignMenuInput {
  menu_id: string;
  /** `null` applies the menu to all seven days. */
  days_of_week: DayOfWeek[] | null;
}
